R/GeospatialFunctions.R

```diff
@@ -0,0 +1,612 @@
+  
+  print("There are no ATTAINS features associated with these WQP observations. Returning original dataframe with empty ATTAINS columns and empty ATTAINS geospatial features.")
+  
+  # return a modified `data` with empty ATTAINS-related columns:
+  no_ATTAINS_data <- data %>%
+    dplyr::mutate("ATTAINS.organizationid" = NA, "ATTAINS.submissionid" = NA, "ATTAINS.hasprotectionplan" = NA,
+                  "ATTAINS.assessmentunitname" = NA, "ATTAINS.nhdplusid" = NA, "ATTAINS.tas303d" = NA,                                                  
+                  "ATTAINS.isthreatened" = NA, "ATTAINS.state" = NA, "ATTAINS.on303dlist" = NA,                                               
+                  "ATTAINS.organizationname" = NA, "ATTAINS.region" = NA, "ATTAINS.Shape_Length" = NA,                                             
+                  "ATTAINS.reportingcycle" = NA, "ATTAINS.assmnt_joinkey" = NA, "ATTAINS.hastmdl" = NA,                                                  
+                  "ATTAINS.orgtype" = NA, "ATTAINS.permid_joinkey" = NA, "ATTAINS.catchmentistribal" = NA,                                        
+                  "ATTAINS.ircategory" = NA, "ATTAINS.waterbodyreportlink" = NA, "ATTAINS.assessmentunitidentifier" = NA,                                 
+                  "ATTAINS.overallstatus" = NA, "ATTAINS.isassessed" = NA, "ATTAINS.isimpaired" = NA,                                               
+                  "ATTAINS.has4bplan" = NA, "ATTAINS.huc12" = NA, "ATTAINS.hasalternativeplan" = NA,                                      
+                  "ATTAINS.visionpriority303d" = NA, "ATTAINS.areasqkm" = NA, "ATTAINS.catchmentareasqkm" = NA,                                       
+                  "ATTAINS.catchmentstatecode" = NA, "ATTAINS.catchmentresolution" = NA, "ATTAINS.Shape_Area" = NA) %>%
+    tibble::rowid_to_column(var = "index")
+  
+  if(return == TRUE){
+  
+  ATTAINS_catchments <- NULL
+  ATTAINS_lines <- NULL
+  ATTAINS_points <- NULL
+  ATTAINS_polygons <- NULL
+  
+  return(list("TADA_with_ATTAINS" = no_ATTAINS_data,
+              "ATTAINS_catchments" = ATTAINS_catchments,
+              "ATTAINS_points" = ATTAINS_points,
+              "ATTAINS_lines" = ATTAINS_lines,
+              "ATTAINS_polygons" = ATTAINS_polygons))
+  } else {
+    
+    return(no_ATTAINS_data)}
+  } else {
+  
+  suppressMessages(suppressWarnings({
+    # ... otherwise link WQP features to the ATTAINS catchment feature(s) they land in:
+    TADA_with_ATTAINS <- TADA_DataRetrieval_data %>%
+      # left join = TRUE to preserve all observations (with or without ATTAINS features):
+      sf::st_join(., nearby_catchments, left = TRUE)
+    
+    # ... otherwise link WQP features to the ATTAINS catchment feature(s) they land in:
+    TADA_data <- TADA_DataRetrieval_data %>%
+      # left join = TRUE to preserve all observations (with or without ATTAINS features):
+      sf::st_join(., nearby_catchments, left = TRUE)
```

Comment:
It looks to me like one of these objects might be redundant, but I'm not 100% sure.